tracer/src/Datadog.Trace/AppSec/BodyExtractor.cs

```diff
@@ -0,0 +1,180 @@
+// <copyright file="BodyExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec
+{
+    internal static class BodyExtractor
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(BodyExtractor));
+
+        private static readonly Regex NameExtractor = new Regex(@"\<(?'PropertyName'[^\>]+)\>", RegexOptions.Compiled);
+
+        private static readonly HashSet<Type> AdditionalPrimitives = new()
+        {
+            typeof(string),
+            typeof(decimal),
+            typeof(Guid),
+            typeof(DateTime),
+            typeof(DateTimeOffset),
+            typeof(TimeSpan)
+        };
+
+        internal static object GetKeysAndValues(object body)
+        {
+            var visted = new HashSet<object>();
+            var item = ExtractType(body.GetType(), body, 0, visted);
+
+            return item;
+        }
+
+        private static bool IsOurKindOfPrimitive(Type t)
+        {
+            return t.IsPrimitive || AdditionalPrimitives.Contains(t) || t.IsEnum;
+        }
+
+        private static void ExtractProperties(Dictionary<string, object> dic, object body, int depth, HashSet<object> visited)
+        {
+            if (visited.Contains(body))
+            {
+                return;
+            }
+
+            visited.Add(body);
+
+            Log.Debug("ExtractProperties - body: {Body}", body);
+
+            var fields =
+                    body.GetType()
+                        .GetFields(BindingFlags.Instance | BindingFlags.NonPublic)
+                        .Where(x => x.IsPrivate && x.Name.EndsWith("__BackingField"))
```

Comment:
It's our intention to only auto properties for the moment. I believe this is safer that alternatives which could create subtle bugs in users programs.

We could get the objects properties and read them via reflection, instead of reading the fields associated with the auto properties, but this would mean that any code in the get property would executed and in some cases this could contain side effects. By just reading the field we are sure we don't affect the application state.

This could mean we don't read data from some properties that would be useful to scan, but I think it's better to start with this low risk approach. Once we've seen this code running production successfully, we'll probably revisit it and see if there's a way we could read more properties or fields safely.